nix/nixos/socketmaster-service.nix

```
@@ -0,0 +1,47 @@
+{ config, lib, systemConfig, ... }:
+let
+  inherit (lib)
+    mkOption
+    types
+    ;
+  inherit (types)
+    str
+    ;
+in
+{
+  options = {
+
+    command = mkOption {
+      description = ''
+        Program to start
+      '';
+      type = str;
+      # no default
+    };
+
+    startMillis = mkOption {
+      description = ''
+        How long the new process takes to boot in milliseconds.
+      '';
+      default = 3000;
+    };
+
+    ### Internal ###
+
+    systemdServiceModule = mkOption {
+      internal = true;
+      # https://github.com/NixOS/nixpkgs/pull/163617
+      type = types.deferredModule or types.raw or types.unspecified;
+    };
+  };
+
+  config = {
+    systemdServiceModule = { ... }: {
+      # Avoid automatic restarts. This will trigger a reload instead.
+      reloadIfChanged = true;
+      serviceConfig.ExecStart = [
+        "${systemConfig.socketmaster.package}/bin/socketmaster -command ${lib.escapeShellArg config.command} -start ${toString config.startMillis} -listen fd://3"
```

```suggestion
        "${systemConfig.socketmaster.package}/bin/socketmaster -command /run/socketmaster/${config.name} -start ${toString config.startMillis} -listen fd://3"
```
If I'm not mistaken the command needs to point to a mutable location so that on reload, it would load the new command.

As an aside, the command script might also include a bunch of env vars.

---

I figured this could be covered by an entry in a config file instead. Not dissimilar to a systemd unit file, but far more limited, because we can't change many of those fancy settings like security or mounts related ones once the systemd service is running.